# Data Exploration – National Corrections Reporting Program (NCRP)

### Dataset Source
National Corrections Reporting Program, 1991–2020: Selected Variables  
(ICPSR 38492)

This dataset provides individual-level prison records across U.S. states over multiple decades, including information on inmate demographics, offense types, admission and release timing, and incarceration characteristics.

---

## Dataset Structure

The dataset is organized into multiple subfolders, each containing a tab-separated data file (`.tsv`) representing a different component of the NCRP data.

- **DS0001** – Inmate-level records including admission, release, and demographic characteristics
- **DS0002** – Records organized by reporting year with inmate and offense characteristics
- **DS0003** – Prison release records with information on release year, release type, and time served
- **DS0004** – prison population at the end of each year*

Each dataset contains structured tab-separated files that can be loaded and explored using Python.

---

## Exploration Goal

Below, we quickly inspect each dataset to understand:

- what information each file contains  
- which variables may be useful for our project  
- which data we still need but may be missing  
- how these datasets could potentially be combined or aggregated for analysis

______________________

**DS-01**

In [ ]:
import pandas as pd

df = pd.read_csv("DS0001/38492-0001-Data.tsv", sep="\t")
df.head()

,ABT_INMATE_ID,SEX,ADMTYPE,OFFGENERAL,EDUCATION,ADMITYR,RELEASEYR,MAND_PRISREL_YEAR,PROJ_PRISREL_YEAR,PARELIG_YEAR,SENTLGTH,OFFDETAIL,RACE,AGEADMIT,AGERELEASE,TIMESRVD,RELTYPE,STATE
0,A012021000000090128,1,1,2,9,2013,2013,9999,9999,9999,4,8,1,5,5,0,3,1
1,A012021000000110168,1,1,1,9,2004,9999,9999,9999,9999,4,3,1,5,9,9,9,1
2,A012021000000090187,1,1,2,9,2009,2009,9999,9999,9999,0,11,9,1,1,0,1,1
3,A012021000000010425,1,3,3,9,2016,2016,9999,9999,9999,2,12,9,3,3,0,3,1
4,A012021000000090117,1,1,1,9,2009,2017,9999,9999,9999,4,4,9,1,2,3,1,1


**Dataset Exploration – DS0001 (Inmate Characteristics)**

Each row in this dataset represents an individual inmate observation.

### Variables

- **ABT_INMATE_ID** – Unique identifier for each inmate
- **SEX** – Sex of the inmate
- **ADMTYPE** – Admission type to prison
- **OFFGENERAL** – General category of offense
- **OFFDETAIL** – Detailed offense category
- **EDUCATION** – Education level of the inmate
- **ADMITYR** – Year the inmate was admitted to prison
- **RELEASEYR** – Year the inmate was released from prison
- **MAND_PRISREL_YEAR** – Mandatory prison release year
- **PROJ_PRISREL_YEAR** – Projected prison release year
- **PARELIG_YEAR** – Year the inmate becomes eligible for parole
- **SENTLGTH** – Length of the sentence
- **RACE** – Race of the inmate
- **AGEADMIT** – Age at admission
- **AGERELEASE** – Age at release
- **TIMESRVD** – Time served in prison
- **RELTYPE** – Type of release from prison
- **STATE** – State where the incarceration occurred


### Purpose of this Dataset

This dataset contains individual-level information on incarcerated individuals, including demographics, offense characteristics, admission and release timing, and sentence details.

Key variables for our analysis include:
- **STATE** – allows aggregation to the state level
- **ADMITYR / RELEASEYR** – allow construction of timelines
- **SENTLGTH and TIMESRVD** – indicate incarceration duration
- **RELTYPE** – may help identify release conditions

This dataset may help us construct measures related to incarceration patterns and prisoner characteristics across states and years, which could be useful when linking prison populations to recidivism outcomes.


**DS-02**

In [ ]:
df2 = pd.read_csv("DS0002/38492-0002-Data.tsv", sep="\t")
df2.head()

/var/folders/rq/6gkwkw8n79d9pw5r3y76nfqc0000gn/T/ipykernel_74136/1027233313.py:1: DtypeWarning: Columns (0: SENTLGTH) have mixed types. Specify dtype option on import or set low_memory=False.
  df2 = pd.read_csv("DS0002/38492-0002-Data.tsv", sep="\t")


,SEX,RPTYEAR,STATE,EDUCATION,ADMTYPE,OFFGENERAL,MAND_PRISREL_YEAR,PROJ_PRISREL_YEAR,PARELIG_YEAR,ADMITYR,SENTLGTH,OFFDETAIL,RACE,AGEADMIT
0,1,1991,1,2,1,2,9999,9999,9999,1991,2,7,2,1
1,2,1991,1,1,1,2,9999,9999,9999,1991,0,8,2,1
2,1,1991,1,9,1,3,9999,9999,9999,1991,2,12,2,2
3,1,1991,1,2,1,4,9999,9999,9999,1991,3,13,1,1
4,1,1991,1,1,1,2,9999,9999,9999,1991,3,8,2,1


**Dataset Exploration – DS0002 (Admissions by Report Year)**

Each row appears to represent an inmate observation associated with a specific **reporting year**.

### Variables

- **SEX** – Sex of the inmate
- **RPTYEAR** – Reporting year of the observation
- **STATE** – State where incarceration occurred
- **EDUCATION** – Education level of the inmate
- **ADMTYPE** – Type of admission
- **OFFGENERAL** – General offense category
- **OFFDETAIL** – Detailed offense category
- **ADMITYR** – Year the inmate was admitted to prison
- **MAND_PRISREL_YEAR** – Mandatory prison release year
- **PROJ_PRISREL_YEAR** – Projected prison release year
- **PARELIG_YEAR** – Parole eligibility year
- **SENTLGTH** – Sentence length
- **RACE** – Race of the inmate
- **AGEADMIT** – Age at admission

**DS-03**

In [ ]:
df3 = pd.read_csv("DS0003/38492-0003-Data.tsv", sep="\t")
df3.head()

/var/folders/rq/6gkwkw8n79d9pw5r3y76nfqc0000gn/T/ipykernel_74136/3728648662.py:1: DtypeWarning: Columns (0: SENTLGTH) have mixed types. Specify dtype option on import or set low_memory=False.
  df3 = pd.read_csv("DS0003/38492-0003-Data.tsv", sep="\t")


,SEX,RPTYEAR,STATE,EDUCATION,ADMTYPE,OFFGENERAL,MAND_PRISREL_YEAR,PROJ_PRISREL_YEAR,PARELIG_YEAR,ADMITYR,SENTLGTH,OFFDETAIL,RACE,AGEADMIT,RELYR,RELTYPE,AGERLSE,TIMESRVD_REL
0,1,1991,1,3,1,1,9999,9999,9999,1983,5,1,1,2,1991,1,3,3
1,1,1991,1,2,1,1,9999,9999,9999,1987,4,4,1,2,1991,2,2,2
2,1,1991,1,2,1,3,9999,9999,9999,1987,4,12,1,4,1991,3,4,2
3,1,1991,1,2,1,2,9999,9999,9999,1989,4,7,2,3,1991,2,3,2
4,1,1991,1,3,1,4,9999,9999,9999,1989,3,13,9,2,1991,2,2,2


**Dataset Exploration – DS0003 (Prison Releases)**

Each row represents an inmate who was **released from prison** in a given reporting year.

### Variables

- **SEX** – Sex of the inmate
- **RPTYEAR** – Reporting year
- **STATE** – State where incarceration occurred
- **EDUCATION** – Education level
- **ADMTYPE** – Admission type
- **OFFGENERAL** – General offense category
- **OFFDETAIL** – Detailed offense category
- **ADMITYR** – Year admitted to prison
- **RELYR** – Year of release
- **RELTYPE** – Type of release (parole, sentence completion, etc.)
- **SENTLGTH** – Length of sentence
- **AGERLSE** – Age at release
- **AGEADMIT** – Age at admission
- **TIMESRVD_REL** – Time served before release
- **RACE** – Race of the inmate
- **MAND_PRISREL_YEAR** – Mandatory prison release year
- **PROJ_PRISREL_YEAR** – Projected prison release year
- **PARELIG_YEAR** – Parole eligibility year

### Notes

This dataset focuses on **prison releases**, which may be important for studying post-release outcomes and potential recidivism patterns.

Key variables for analysis include:
- **STATE**
- **RPTYEAR / RELYR**
- **TIMESRVD_REL**
- **RELTYPE**

These variables may help construct state-year measures of prison releases and incarceration durations.

**DS-04**

In [ ]:
df4 = pd.read_csv("DS0004/38492-0004-Data.tsv", sep="\t")
df4.head()

,SEX,RPTYEAR,STATE,EDUCATION,ADMTYPE,OFFGENERAL,MAND_PRISREL_YEAR,PROJ_PRISREL_YEAR,PARELIG_YEAR,ADMITYR,SENTLGTH,OFFDETAIL,RACE,AGEADMIT,AGEYREND,TIMESRVD_YREND
0,1,1999,6,9,1,1,9999,9999,9999,1955,6,1,2,1,5,4
1,1,1999,6,9,1,1,9999,9999,9999,1955,6,1,4,2,5,4
2,1,1999,6,9,1,1,9999,9999,9999,1957,6,5,2,1,5,4
3,1,1999,6,9,1,1,9999,9999,9999,1957,6,1,1,1,5,4
4,1,1999,6,9,1,1,9999,9999,9999,1957,6,1,1,2,5,4


**Dataset Exploration – DS0004 (Year-End Prison Population)**

Each row appears to represent an inmate who was **incarcerated at the end of a reporting year**.

### Variables

- **SEX** – Sex of the inmate  
- **RPTYEAR** – Reporting year  
- **STATE** – State where incarceration occurred  
- **EDUCATION** – Education level  
- **ADMTYPE** – Admission type  
- **OFFGENERAL** – General offense category  
- **OFFDETAIL** – Detailed offense category  
- **ADMITYR** – Year admitted to prison  
- **SENTLGTH** – Sentence length  
- **AGEADMIT** – Age at admission  
- **AGEYREND** – Age at the end of the reporting year  
- **TIMESRVD_YREND** – Time served as of the end of the reporting year  
- **RACE** – Race of the inmate  
- **MAND_PRISREL_YEAR** – Mandatory prison release year  
- **PROJ_PRISREL_YEAR** – Projected prison release year  
- **PARELIG_YEAR** – Parole eligibility year


### Notes

This dataset appears to capture the **prison population at the end of each year**.

Key useful variables:
- **STATE**
- **RPTYEAR**
- **TIMESRVD_YREND**

This dataset could allow construction of **state-year prison population or incarceration duration measures**, which may be useful when studying incarceration patterns across states and time.

______________________________________


## Dataset Choice

**DS0001** is the most useful dataset.

It contains a **unique inmate ID (`ABT_INMATE_ID`)**, which allows individuals to be tracked across records. It also includes key timeline variables such as **ADMITYR**, **RELEASEYR**, and **STATE**.

The other datasets lack a unique identifier and mainly describe admissions, releases, or year-end prison populations.

______

## DS0001 – Basic Data Inspection

In [ ]:
df01 = df
df01["ADMITYR"].min(), df01["ADMITYR"].max()

(np.int64(1950), np.int64(9999))

In [ ]:
df01["RELEASEYR"].min(), df01["RELEASEYR"].max()

(np.int64(1971), np.int64(9999))

In [ ]:
df01.shape

(13897856, 18)

In [ ]:
df01["ABT_INMATE_ID"].nunique()

7316619

In [ ]:
df01["ABT_INMATE_ID"].duplicated().sum()

np.int64(6581237)

In [ ]:
df01["STATE"].nunique()

44

In [ ]:
df01.isna().sum()

ABT_INMATE_ID        0
SEX                  0
ADMTYPE              0
OFFGENERAL           0
EDUCATION            0
ADMITYR              0
RELEASEYR            0
MAND_PRISREL_YEAR    0
PROJ_PRISREL_YEAR    0
PARELIG_YEAR         0
SENTLGTH             0
OFFDETAIL            0
RACE                 0
AGEADMIT             0
AGERELEASE           0
TIMESRVD             0
RELTYPE              0
STATE                0
dtype: int64

### DS0001 – Summary

- **Observations:** 13,897,856
- **Unique inmates:** 7,316,619
- **Repeat inmate records:** 6,581,237
- **States covered:** 44
- **Missing values:** none

**Admission years:** start at 1950  
**Release years:** start at 1971  

The value **9999 appears to represent cases where release has not yet occurred or is not recorded.**

The presence of repeated inmate IDs suggests that individuals can appear multiple times in the dataset, which may allow identification of **repeat incarcerations**.

In [ ]:
df01["ABT_INMATE_ID"].value_counts().head()

ABT_INMATE_ID
A062021000000022552    46
A362021000000005215    46
A062021000000058980    40
A062021000000077121    39
A062021000000041577    39
Name: count, dtype: int64

**Key EDA Questions:**
- Which states provide data consistently?
- What cleaning is needed for null values like 9999?
- How can we best operationalize (measure) recidivism by counting repeating ID records?